Add design proposal for better obsoletion (#1)
terrajobst/designs: accepted/better-obsoletion/better-obsoletion.md

# Better Obsoletion

**PM** [Immo Landwerth](https://github.com/terrajobst)

No framework, no matter how carefully crafted, is immune to making mistakes. For
a framework to last the test of time it's vital that it can be evolved to enable
developers to build the kind of applications that matter to their business,

> which is an ever changing landscape and often puts additional constraints on the

Reviewer: ```suggestion
which is an ever-changing landscape and often puts additional constraints on the
```

developer's code that didn't exist before.

The approach to productivity in .NET is called "the pit of success", which is
the idea that developers will more or less stumble their way via code completion
towards the general vicinity of a correct solution. However, as the platform
grows and ages, the amount of mistakes in the platform also grows. And more
often than not the API that looks the most promising might be the wrong one to
use. So it becomes important that we can point developers the right way.

Since .NET Framework 1.0, we were able to mark APIs as obsoleted via a built-in

> attribute. Any usage of such an API results in the compiler emitting a warning

Reviewer 2: Consider "diagnostic" instead of "warning". The compiler can emit a warning or error here.

Author: Good call, replaced

Obsoletion via the `[Obsolete]` attribute works pretty well when: only a few
APIs are affected, it's easy to remove their usage, and guidance can be provided
as a short one-line string.

In practice these requirements are often not satisfiable:

* **Technologies have large API surfaces**. Obsoleting an entire technology
  means that we'll have to annotate all entry points to that technology, which
  often means obsoleting a *set of APIs*, such as types and static members.

* **Removing obsoleted APIs isn't always practical**. It would be nice if we
  could always remove dependencies on obsoleted technologies but this isn't
  always possible if they are used heavily in existing code bases. In those
  cases, you typically want a way to ignore obsoletion warnings regarding
  specific technologies but not others so that you don't accidentally start
  depending on obsoleted technologies when you're evolving your code base.

* **Guidance is complex**. Obsoleted APIs rarely have a drop-in replacement.
  After all, we don't obsolete based on their naming or shape but usually due to
  fundamental design issues. That means the replacements are typically a nuanced

>   choice and require at least a page of explanation, rather than a one-liner. 

Reviewer 2: I think it's okay if you say `AppDomain` here

:trollface:

The component is designed for a traditional two-tier environment and includes
data access that handles retrieving and storing business objects from and to SQL
Server. Abigail is now tasked to refactor that code to use REST-based services
instead so that the application can eventually be converted to a cloud-based
offering.

Abigail would like to use the brand new JSON APIs so she updates the project
from .NET Core 2.1 to .NET Core 3.1. However, after upgrading the notices that
we obsoleted non-generic collections and she gets a bunch of warnings against
her code:

| Code   | Description              | Project                        |
|--------|--------------------------|--------------------------------|
| DE0006 | 'ArrayList' is obsolete  | Fabrikam.Billing.BusinessLogic |
| DE0006 | 'Hashtable' is obsolete  | Fabrikam.Billing.BusinessLogic |

Due to the heavy use, there are dozens of warnings like this throughout her
code, so she doesn't want to use `#pragma warning disable` for each affected
call site. Instead, she decides to suppress all obsoletion warnings for
non-generic collections by adding the ID `DE0006` to the

> `Fabrikam.Billing.BusinessLogic` project's `NoWarn` list.

Reviewer 2: This doesn't seem to cover what happens when we actually delete these APIs out from under Abigail. Is that deliberate? Not sure the value of `Obsolete` if we never delete the APIs.

Author: That's deliberate because I don't foresee us (BCL) being able to remove APIs ever. That doesn't mean that higher levels can't (or shouldn't); it just means that obsoleting in and of itself adds value even if we can't remove the API. I've added a section to talk about this in more detail.

## More Details

Bill is an architect and maintains the authentication and authorization related
components in Fabrikam Billing. Based on guidance from 2005, the code uses
`SecureString`.

Bill is now tasked with refactoring this code to support Azure Active Directory
based authentication. Since most of the code base is now being upgraded to .NET
Core 3.1, he starts by retargeting the project. After doing so, he gets a bunch
of obsoletion warnings:

| Code   | Description                 | Project               |
|--------|-----------------------------|-----------------------|
| DE0001 | 'SecureString' is obsolete  | Fabrikam.Billing.Auth |

Trying to understand why, he clicks the warning ID `DE0001` in the error list
which takes him to a documentation page explaining the issues with
`SecureString`. There he learns that this type doesn't work well for
cross-platform scenarios due to the lack of automatic encryption facilities and
the recommendation is to stop using it altogether. Bill knows that moving to the
cloud also means they are considering moving to Linux, so he takes note that
they also need to plan for removing their dependency on `SecureString`.

## Requirements

> ### Goals

Reviewer 2: Is there a reason providing code fixes is not a goal here?

Author: I should list it under non-goals. The reason being that it's unlikely that we can provide fixers for 90% of the APIs we'd like to obsolete:

* **SecureString**. The fix is to not use passwords to begin with. Just making the `Password` property of type `string` isn't what you want.

* **Non-generic collections**. You can't just replace the types (first, generic inference would be hard) but also you've to update usage patterns that assume `Hashtable` returns `null` for missing keys.

* **Remoting**. Requires moving to some other IPC technology. The code needs to be written very differently.

There might be some cases where we can, but to me this wouldn't be required to call success. My goal is educating people and preventing new code to depend on obsoleted tech.

* **Global suppressions are specific**. Developers must be able to suppress
  obsoletion warnings for all obsoletion warning that relate to the same
  technology, rather than having to turn all obsoletion warnings off or having
  to resort to suppress every single call site individually.

* **Documentation pages are easily accessible**. Developers must be able to go
  from the warning/error in the error list straight to the documentation that

>   provides the details.

Reviewer 2: Is there a reason we don't have deleting APIs as a goal?

Author: Clarified.

### Non-Goals

* **Detecting specific usage patterns**. Sometimes, an API isn't obsoleted, just
  a particular combination of arguments or specific values. If we need to
  identify these, we'll author a custom analyzer.

* **Being able to obsolete APIs in already shipped versions of .NET**. The
  attribute must be applied to the API surface that the compiler sees and thus
  must be present in the reference assemblies. We can only obsolete in new

>   versions of the platform. In other words, this doesn't support a side car file

Reviewer: ```suggestion
  versions of the platform. In other words, this doesn't support a sidecar file
```

  that can be applied to an existing version of .NET.

## Design

> The design proposal piggy backs on the existing `ObsoleteAttribute` that is

Reviewer: ```suggestion
The design proposal piggybacks on the existing `ObsoleteAttribute` that is
```

> already supported by all .NET compiler. The idea is to expand this a little to

Reviewer: ```suggestion
already supported by all .NET compilers. The idea is to expand this a little to
```

support the desired scenarios. Alternatives are listed afterwards.

### API Changes

We'd add the following properties to `ObsoleteAttribute`:

```C#
namespace System
{
    public sealed class ObsoleteAttribute : Attribute
    {
        // Existing:
        //
        // public ObsoleteAttribute();
        // public ObsoleteAttribute(string message);

>         // public ObsoleteAttribute(string message, bool error);

Reviewer 3: Worth adding another constructor that additionally takes `diagnosticId` and `urlFormat`?

Author: I thought about that but it would also have to take a message in order to avoid confusion between the diagnostic ID and the message; however, I think we generally don't want to put a message when using a URL.

        // public bool IsError { get; }
        // public string Message { get; }

        // New:
        public string DiagnosticId { get; set; }
        public string UrlFormat { get; set; }
        public string Url => UrlFormat == null ? null : string.Format(UrlFormat, DiagnosticId);
    }
}
```

Instead of taking the URL directly, the API takes a format string. This allows
having a generic URL that includes the diagnostic ID. This avoids having to
repeat the ID twice and thus making a copy & paste mistake.

### Compiler Changes

The compilers would be enlightened about these new properties and use them as
follows:

> * If `DiagnosticId` is `null`, use the existing diagnostic ID (e.g. `CS618` in

Reviewer: ```suggestion
* If `DiagnosticId` is `null`, use the existing diagnostic ID (e.g. `CS0618` in
```

  C#). Otherwise, use the specified warning ID.

> * If `Url` is not `null`, use it as the diagnostic link when rendering them in

Reviewer 3: Is a non-null `Url` only used if `DiagnosticId` is also non-null?

Author: No, I'd treat URL and ID as orthogonal features. Both can be used, either, and neither.

  the IDE.

### Samples

Here is how the attribute would be applied:

```C#
namespace System.Collections
{

>     [Obsolete(DiagnosticId = "DE0006", UrlFormat="http://aka.ms/obsolete/{0}")]

Reviewer 3: Do we already have guidance, e.g. from Roslyn analyzers, on how a user picks a good diagnostic id?

Author: Good point I've added an open issue to the doc

    public class ArrayList
    {
        // ...
    }

>     [Obsolete(DiagnosticId = "DE0006", UrlFormat="http://aka.ms/obsolete/{0}")]

Reviewer: Presumably this UrlFormat would come from some other const, rather than a literal (thus justifying its existence as a format string)

Author: Yep

  C#). Otherwise, use the specified warning ID.
* If `Url` is not `null`, use it as the diagnostic link when rendering them in
  the IDE.
* The compiler should assume that `Url` and `DiagnosticId` are independent
  features, in other words both can be used, either, and neither.

### Samples

Here is how the attribute would be applied:

```C#
internal static class Obsolete
{
    public const string Url = "http://aka.ms/obsolete/{0}";

    // NOTE: Keep this sorted by the the ID's value,
    //       not the name of the field!
    public const string SecureStringId = "BCL0001";
    public const string NonGenericCollectionsId = "BCL0006";
}

namespace System.Collections
{
    [Obsolete(DiagnosticId = Obsolete.NonGenericCollectionsId, UrlFormat=Obsolete.Url)]
    public class ArrayList
    {
        // ...
    }

    [Obsolete(DiagnosticId = Obsolete.NonGenericCollectionsId, UrlFormat=Obsolete.Url)]
    public class Hashtable
    {
        // ...
    }
}

namespace System.Security
{
    [Obsolete(DiagnosticId = Obsolete.SecureStringId, UrlFormat=Obsolete.Url)]
    public sealed class SecureString
    {
        // ...
    }
}

> ```

Reviewer 4: I see you changed the sample a bit since I left my [other comment](https://github.com/dotnet/designs/pull/60#issuecomment-479606367). 😉

I still feel like having a special attribute to define defaults is the easier solution. It removes the extra noise of having `UrlFormat=Obsolete.Url` all over the place. Additionally, (until we get first-class code generators in 20XX) it's easier to inject an additional attribute from the build system than to modify an existing constant.

It's maybe worth mentioning that this isn't a new pattern. [DllImportAttribute.CharSet](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute.charset) has [DefaultCharSetAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.defaultcharsetattribute). (Although on the contrary, [I don't think this attribute is well known](https://github.com/search?l=C%23&q=%22module%3A+DefaultCharSet%22&type=Code).)

Author: Maybe. The compiler team has a very strong opinion hat ambient settings that influence what a particular line of code means is bad. So in that sense, the code:

```C#
[Obsolete(DiagnosticId = "XX123")]
```

Should mean that no URL is being used. It's problematic if it sometimes means no URL and sometimes we use a URL, depending on some attribute defined in some other file. Personally, I don't care for the duplication because these attributes should be sparingly used. If we sprinkle them all over the code, we've got a problem.

> It's maybe worth mentioning that this isn't a new pattern.

And this pattern, in conjunction with others (such as checked/unchecked) have proven to be the cause for many bugs when moving code between projects. Past mistakes don't mean we need to repeat them ;-)

/cc @jaredpar

Reviewer 4: Yeah, I can definitely agree with all of that. Especially when the setting is defined in some other file.

I wasn't properly considering that scenario. I thought of `DefaultObsoleteUrl` missing entirely (which would be annoying but not misleading), but not when you're suddenly using the *wrong* URL. Getting a link to `https://aka.ms/ARES0123` instead of `https://pgn.moe/ARES0123` would be confusing and annoying.

***OPEN ISSUE**: We need to decide on guidance to avoid ID conflicts with the
community. We should talk to the Roslyn analyzer team for best practices. Stake
in the ground could be that we should a prefix like `BCL` that is owned by a
single party and responsible to avoid duplicated numbers. Other parties would

> need to choose a different prefix.*

Reviewer 4: This probably gets too complex too quickly, and I imagine it'd be considered too much effort for something that isn't all that common of an issue. Either way, here it is:

One possibility for this is to have diagnostics defined as special types which are then scoped to a namespace.

For example:

```csharp
namespace System.Collections
{
    public sealed class DE0006 : System.DiagnosticId { }

    [Obsolete<DE0006>] // Dreaming csharplang#124 is completed. [Obsolete(typeof(DE0006))] otherwise.
    public class ArrayList
    {
        // ...
    }
}

namespace SomeOtherLibrary
{
    public sealed class DE0006 : System.DiagnosticId { }

    [Obsolete<DE0006>]
    public class SomeOldClass { }
}
```

Examples for how to suppress the diagnostic:

```csharp
using System.Collections;

#pragma warning disable DE0006

public class MyLegacyClass
{
    public ArrayList OldList;
}
```

```csharp
using System.Collections;
using SomeOtherLibrary;

//#pragma warning disable DE0006 // CS0104: 'DE0006' is an ambiguous reference between 'System.Collections.DE0006' and 'SomeOtherLibrary.DE0006'
#pragma warning disable System.Collections.DE0006

public class MyOtherLegacyClass
{
    public ArrayList OldList;
    public SomeOldClass OldClass; // Warning SomeOtherLibrary.DE0006: SomeOldClass is deprecated!
}
```

Reviewer 4: I forgot to point out, the main reason for the use of a type to mark diagnostic IDs is that it makes it easier to scope it to a namespace. (My original thought was having attributes on the namespace that define what diagnostic IDs exist within it, but you can't put attributes on namespaces. Another perhaps more sane alternative would be to have the static `Obsolete` class be a formalized thing that is recognized by the compiler, then you just have those in each relevant namespace.)

Also, in the magical world that generic attribute are somehow added before this feature, `Obsolete<T>` looks like this:

```csharp
[AttributeUsage(/* same as ObsoleteAttribute */)]
public sealed class Obsolete<T> : Attribute
    where T : DiagnosticId
{ }
```

As an added benefit, this theoretical `DiagnosticId` type could provide extra things like the compiler message, the URL, and whether it's a compilation error or just a warning.

Author: I don't think this solves many issue. It's in the realm of "cutesy" APIs, i.e. it looks like you're trying too hard to make it look elegant without expanding the feature's versatility. Besides, attributes can't use generics.

I do, however, like the idea of namespace qualifying the ID but this can be easily done as a string constant. We don't need to define the IDs as types.

Reviewer 4: > It's in the realm of "cutesy" APIs,

I did try to imply that it was a bit over the top. Although the idea of using the special types was to address specific issues I thought of with namespaced string diagnostics.

> Besides, attributes can't use generics.

Hence my comment about https://github.com/dotnet/csharplang/issues/124

> I do, however, like the idea of namespace qualifying the ID but this can be easily done as a string constant. We don't need to define the IDs as types.

I did try to clarify that was my original idea in my second comment (which I think I might've added after you loaded the page - although I could've done better with that comment either way.)

My main concern would be that if you're using the namespace of the deprecated type, you can't have a diagnostic that spans multiple namespaces. If you're saying that the namespace would be part of the string provided to the `ObsoleteAttribute` definition, that would be less of a problem.

Part of the idea behind using special types for the different diagnostics was to leverage the existing ability for the compiler to resolve types and provide errors about ambiguous ones. It also gives you a reliable way to see everywhere the diagnostic is used/suppressed with Find All References.

Author: > My main concern would be that if you're using the namespace of the deprecated type, you can't have a diagnostic that spans multiple namespaces.

That's true, although in practice you only need a prefix for the owner of the API, which could be root (e.g. the company name or organization). For us, good examples would `System` and `Microsoft.AspNetCore`.

Reviewer 4: To clarify, I was referring to things being problematic if the namespace was inferred from the type's namespace. IE:

```csharp
namespace System.Collections
{
    // "Fully qualified" diagnostic is System.Collections.BCL0006
    [Obsolete(DiagnosticId = "BCL0006")]
    public class ArrayList
    { }
}

namespace System.Collections.Specialized
{
    // "Full qualified" diagnostic is System.Collections.Specialized.BCL0006
    [Obsolete(DiagnosticId = "BCL0006")]
    public class HybridDictionary
    { }
}
```

If you're wanting to make the namespace part of the string (IE: `[Obsolete(DiagnosticId = "System.Collections.BCL0006")]`), then that isn't an issue anymore. The new issue with that would be making sure that people actually put that prefix in. (It could just be documented as a best practice and just hope most people do it.)

in the attribute's message property, such as:

```C#
namespace System.Collections
{
    [Obsolete("::BCL0006|http://aka.ms/obsolete/{0}")]
    public class ArrayList
    {
        // ...
    }
}
```

The benefit is that it doesn't require new API and that that the call sites
are slightly more compact.

The downside is more cryptic use and weak contract between the compiler and the
API surface, which might also result in accidental recognition.

#### Separate Attribute

We could also introduce a new attribute and move detection to an analyzer that
is referenced by default. The shape could look as follows:

```C#
namespace System
{
    public sealed class DeprecatedAttribute : Attribute
    {
        public DeprecatedAttribute(string diagnosticId);
        public DeprecatedAttribute(string diagnosticId, string urlFormat);

        public string DiagnosticId { get; set; }
        public string UrlFormat { get; set; }
        public string Url => UrlFormat == null ? null : string.Format(UrlFormat, DiagnosticId);
    }
}
```

The irony is that this would effectively mean that we're obsoleting the
`ObsoleteAttribute` which now begs the question whether we should mark it

> `[Obsolete]` or `[Deprecated]`.

Reviewer 4: EDIT: This comment was originally going to be against deprecating `ObsoleteAttribute` if this was done, but I talked myself out of it. I should've moved it to the above code block instead.

I think having the message in addition to the diagnostic ID is still valuable. It still allows for having the quick one-line description of the problem without forcing the user to go through documentation.

-------------

If you do go with a separate attribute (and it has a message), I know I personally wouldn't mind having it cover other non-recommended uses besides deprecation.

For example, we currently misuse `ObsoleteAttribute` to guard against the usage of prototyping-only APIs in release builds. (It's not a common thing, but generally these APIs let you get at the internal state of a class. Usually to work around a limitation of the implementation until a more proper solution is determined. For example:

```csharp
internal DeviceContext Context { get; }

[EditorBrowsable(EditorBrowsableState.Never)]
#if !DEBUG
[Obsolete("This method is for prototyping purposes only!")]
#endif
public DeviceContext UnsafeGetContext()
    => Context;
```

Ideally we'd use a Roslyn analyzer for this, but this is the quick-and-easy "temporary" (since 2017 in this example) solution we're using. I definitely wouldn't mind something like `[UseDiscouraged("ARES0123", "This method is for prototyping purposes only!")]` instead.

The obvious downside of something more generic like this is all of the sudden the compiler can't make a reasonable default message with no specific one provided.

Author: Hmm, I think we want the meaning to be quite specific. In that sense, I think you'd be better off using what you do today or go with a custom analyzer. If anything, you're convincing me more that evolving `ObsoleteAttribute` is better than going with a new attribute.

Reviewer 4: It's maybe worth pointing out that if `ObsoleteAttribute` is deprecated, its existence mentioned in CLI specification. (Partition II, 21.2.2, Custom attributes defined by the CLS. [That's page 119 in the Microsoft implementation-specific version on this page.](https://visualstudio.microsoft.com/license-terms/ecma-c-common-language-infrastructure-standards/))

Author: Yeah, but I don't think this would stop us though. The spec isn't gospel. ECMA also mentions a ton of APIs, especially around CAS and the old XML Dom, both of which we consider deprecated/obsolete. In the end, we're looking at this from the "would this improve .NET angle". The reason deviation from the spec is problematic is because most deviations also tend to be breaking changes and those are usually not worth it. Nonetheless, we have (and continue to) deviate from the spec if necessary.

Reviewer 4: >Hmm, I think we want the meaning to be quite specific.

That's definitely the easier route. I just thought it should be considered if you're going to go through the effort of adding something completely new.

Reviewer 4: > Nonetheless, we have (and continue to) deviate from the spec if necessary.

Roger that, I just wanted to point it out in case it was something you all cared about.
